ratatui-core/src/symbols/merge.rs

```diff
@@ -0,0 +1,431 @@
+    ///
+    /// Example: `┐` and `┗` will be merged into `╄`
+    Exact,
+
+    /// Merges symbols even if an exact composite unicode character doesn't exist,
+    /// using the closest match.
+    ///
+    /// Example: `╮` and `└` will be merged into `┼`
+    Fuzzy,
+}
+
+impl MergeStrategy {
+    /// Merges two symbols using this merge strategy.
+    pub fn merge<'a>(self, prev: &'a str, next: &'a str) -> &'a str {
+        let (Ok(prev_symbol), Ok(next_symbol)) =
+            (BorderSymbol::from_str(prev), BorderSymbol::from_str(next))
+        else {
+            return next;
+        };
+        if let Ok(merged) = prev_symbol.merge(next_symbol, self).try_into() {
+            return merged;
+        }
+        next
+    }
+}
+
+/// Represents a composite border symbol using individual line components.
+#[derive(Debug, Copy, Clone)]
+struct BorderSymbol {
+    pub right: LineStyle,
+    pub up: LineStyle,
+    pub left: LineStyle,
+    pub down: LineStyle,
+}
+
+impl BorderSymbol {
+    /// Creates a new [`BorderSymbol`], based on individual line styles.
+    #[must_use]
+    pub const fn new(right: LineStyle, up: LineStyle, left: LineStyle, down: LineStyle) -> Self {
+        Self {
+            right,
+            up,
+            left,
+            down,
+        }
+    }
+
+    /// Finds the closest representation of the [`BorderSymbol`], that has a corresponding unicode
+    /// character.
+    #[must_use]
+    pub fn fuzzy(mut self) -> Self {
+        #[allow(clippy::enum_glob_use)]
+        use LineStyle::*;
+
+        // Dashes only include vertical and horizontal lines.
+        if !self.is_straight() {
+            self = self
+                .replace(DoubleDash, Plain)
+                .replace(TripleDash, Plain)
+                .replace(QuadrupleDash, Plain)
+                .replace(DoubleDashThick, Thick)
+                .replace(TripleDashThick, Thick)
+                .replace(QuadrupleDashThick, Thick);
+        }
+
+        // Rounded has only corner variants.
+        if !self.is_corner() {
+            self = self.replace(Rounded, Plain);
+        }
+
+        // There are no Double + Thick variants.
+        // Additionally, some Plain + Double variants don't exist.
+        // TODO: check if this is good enough after adding tests
+        if TryInto::<&str>::try_into(self).is_err() {
+            self = self.replace(Double, Thick);
+        }
+        self
+    }
+
+    /// Return true only if the symbol is a line and both parts have the same [`LineStyle`].
+    fn is_straight(self) -> bool {
+        use LineStyle::Nothing;
+        (self.up == self.down && self.left == self.right)
+            && (self.up == Nothing || self.left == Nothing)
+    }
+
+    /// Return true only if the symbol is a corner and both parts have the same [`LineStyle`].
+    fn is_corner(self) -> bool {
+        use LineStyle::Nothing;
+        // TODO: maybe this can be simplified?
```

Comment:
I'm okay with this at its current state, simplifying it might make it less readable imo